### ISSUE TITLE
Rewrite Cabal in Shake

### ISSUE THREAD

**Issue by ezyang** [haskell/cabal#4174]
The Cabal build system today is very bad at avoiding redundant work. For example, when you ask it for a rebuild, it must always rerun the preprocessors... because it doesn't know any better. cabal new-build has a complicated change tracking mechanism built on top of the old-style Setup interface, because the Cabal library is simply not setup to handle it. Sharing a common, dependency graph creation abstraction, i.e., Shake, will help us solve these problems.

Here are the major architectural points that make sense to me at this point:

* The Shake build system would live parallel to the usual one, NOT in the Cabal library (so that Cabal can be boot without pulling in Shake). There should be a way to ask, e.g., new-build to swap processing of all Simple (and perhaps others) build scripts to use the new system. We probably need to preserve the old system for Custom as it will probably be very hard to maintain library compatibility in the new world order.

* Hadrian has some good ideas about how to design build systems for Haskell code (https://www.staff.ncl.ac.uk/andrey.mokhov/Hadrian.pdf); a rewrite of Cabal in Shake should take advantage of those ideas. (In particular, they have good ways of dealing with the tons and tons of flags we need to pass to the subprograms we call.)

* (More here)

CC @Ericson2314

**Comment by gbaz** [comment]
"so that Cabal can be boot without pulling in Shake"

Skimming the shake deps, it looks like what it depends on, while a bit broader than what cabal does, is actually a pretty minimal set. The nonstandard deps don't have any further non-standard second-order deps themselves. I don't know the current story with cabal deps, but my sense is they've loosened up a tiny bit. In such a circumstance, the broader shake set of deps may not be that bad...

**Comment by phadej** [comment]
`Cabal` is boot library, so extra dependencies are not-so-easy. Luckily Cabal is upgradable package, but still, situation is difficult.

I don't like "Live in parallel". Keeping them comparable is non-trivial work.

I'm not against this task, but would like it to be done when `new-build` fully replaces `build`. If the large drawback atm is "we must rerun preprocessors", then IMHO there isn't urgent need to accomplish this.

**Comment by gbaz** [comment]
ah, right, its cabal-install which can acquire deps (slightly) more freely, i suppose.

**Comment by ezyang** [comment]
See #4175 for another example of something that probably would get better with a Shake based build system.

**Comment by ezyang** [comment]
I mean, "we must always rerun preprocessors" was just one example of many, many operations that we have to redo when we call Cabal. Other examples are the 'configure' steps: whenever you reconfigure (e.g., Cabal file changed) everything has to be redone, even though most of it doesn't change; Shake would help with that as well.

**Comment by Ericson2314** [comment]
@phadej I imagine both cabal-install and ghc-pkg need a not-too-dissimilar subset of Cabal, mainly type definitions and parsing/serializing/marshalling? Perhaps the same, largely self-contained "rump Cabal" could be kept around for them both.

**Comment by ezyang** [comment]
Yes. Similar with Stack; they only need the parsing code, not the build system.

**Comment by Ericson2314** [comment]
Ah, cause they always do external Setup.hs?

**Comment by ezyang** [comment]
Yep.

**Comment by dmwit** [comment]
Another thing that may be worth pondering: shake has a concept of "resources" which are used to limit parallelism. The canonical example is to have just one "linker" resource, because linkers are traditionally very memory hungry and having multiple run at a time can be a recipe for OOM death. To what extent do we want to identify a core set of resources which cabal knows about and manages *across* packages -- so that, e.g., we don't have five packages all in the linking phase at one time?

**Comment by 23Skidoo** [comment]
I have an [old branch](https://github.com/23Skidoo/cabal/tree/num-linker-jobs) that implements limiting the number of linking steps that can run concurrently. Maybe we should dust it off and merge.
